micropip/_commands/freeze.py

```diff
@@ -21,41 +23,55 @@ def freeze() -> str:
     You can use your custom lock file by passing an appropriate url to the
     ``lockFileURL`` of :js:func:`~globalThis.loadPyodide`.
     """
-    packages = deepcopy(REPODATA_PACKAGES)
-    for dist in importlib.metadata.distributions():
-        name = dist.name
-        version = dist.version
-        url = dist.read_text("PYODIDE_URL")
-        if url is None:
-            continue
-
-        sha256 = dist.read_text("PYODIDE_SHA256")
-        assert sha256
-        imports = (dist.read_text("top_level.txt") or "").split()
-        requires = dist.read_text("PYODIDE_REQUIRES")
-        if not requires:
-            fix_package_dependencies(name)
-            requires = dist.read_text("PYODIDE_REQUIRES")
-        if requires:
-            depends = json.loads(requires)
-        else:
-            depends = []
-
-        pkg_entry: dict[str, Any] = dict(
-            name=name,
-            version=version,
-            file_name=url,
-            install_dir="site",
-            sha256=sha256,
-            imports=imports,
-            depends=depends,
-        )
-        packages[canonicalize_name(name)] = pkg_entry
+    return json.dumps(freeze_data())
+
+
+def freeze_data() -> dict[str, Any]:
+    pyodide_packages = deepcopy(REPODATA_PACKAGES)
+    pip_packages = load_pip_packages()
+    package_items = itertools.chain(pyodide_packages.items(), pip_packages)
 
     # Sort
-    packages = dict(sorted(packages.items()))
-    package_data = {
+    packages = dict(sorted(package_items))
+    return {
         "info": REPODATA_INFO,
         "packages": packages,
     }
-    return json.dumps(package_data)
+
+
+def load_pip_packages() -> Iterator[tuple[str, dict[str, Any]]]:
+    return map(
+        package_item,
+        filter(is_valid, map(load_pip_package, importlib.metadata.distributions())),
+    )
+
+
+def package_item(entry: dict[str, Any]) -> tuple[str, dict[str, Any]]:
+    return canonicalize_name(entry["name"]), entry
+
+
+def is_valid(entry: dict[str, Any]) -> bool:
+    return entry["url"] is not None
```

Comment:
```suggestion
    return entry.get("url", None) is not None
```

Reply:
@jaraco if we do this change we can merge it. I'll rebase and apply this and then do a rebase merge.

Reply:
Oh, whoops. This was a mistake. The proper key should be `"file_name"`. Specifically,

```
return entry["file_name"] is not None
```

Since `load_pip_package` sets `file_name` unconditionally, calling `.get()` is unnecessary; the code can rely on that key being present, but it may be set to `None`.

With the patch as suggested, `entry.get("url", None)` will always be `None` and no entry will ever be valid.

Reply:
Oh, and it appears you recognized that in the rewritten commit. Thanks!

Reply:
Yeah, thankfully we have test coverage for this =)